src/main/host/descriptor/socket/inet/tcp.rs

```diff
@@ -168,32 +468,261 @@ impl TcpSocket {
+                    net_ns,
+                    rng,
+                )?;
+
+                Ok::<_, SyscallError>(Some(handle))
+            };
+            socket_ref.with_tcp_state(cb_queue, |state| state.listen(backlog, associate_fn))
+        };
+
+        let handle = match rv {
+            Ok(x) => x,
+            Err(tcp::ListenError::InvalidState) => return Err(Errno::EINVAL.into()),
+            Err(tcp::ListenError::FailedAssociation(e)) => return Err(e),
+        };
+
+        // the `associate_fn` may or may not have run, so `handle` may or may not be set
+        if let Some(handle) = handle {
+            assert!(socket_ref.association.is_none());
+            socket_ref.association = Some(handle);
+        }
+
+        Ok(())
     }
 
     pub fn connect(
-        _socket: &Arc<AtomicRefCell<Self>>,
-        _peer_addr: &SockaddrStorage,
-        _net_ns: &NetworkNamespace,
-        _rng: impl rand::Rng,
-        _cb_queue: &mut CallbackQueue,
+        socket: &Arc<AtomicRefCell<Self>>,
+        peer_addr: &SockaddrStorage,
+        net_ns: &NetworkNamespace,
+        rng: impl rand::Rng,
+        cb_queue: &mut CallbackQueue,
     ) -> Result<(), SyscallError> {
-        todo!();
+        let socket_ref = &mut *socket.borrow_mut();
+
+        // if there was an asynchronous error, return it
+        if let Some(error) = socket_ref.with_tcp_state(cb_queue, |state| state.clear_error()) {
+            // by returning this error, we're probably (but not necessarily) returning a previous
+            // connect() result
+            socket_ref.connect_result_is_pending = false;
+
+            return Err(match error {
+                // TODO: not sure what to return here
+                tcp::TcpError::ResetSent => Errno::EINVAL,
+                tcp::TcpError::ResetReceived => Errno::ECONNREFUSED,
+                tcp::TcpError::TimedOut => Errno::ETIMEDOUT,
+            }
+            .into());
+        }
+
+        // if connect() had previously been called (either blocking or non-blocking), we need to
+        // return the result
+        if socket_ref.connect_result_is_pending {
+            // ignore all connect arguments and just check if we've connected
+
+            // if the ESTABLISHED flag isn't set, then it's still connecting
+            if !socket_ref
+                .tcp_state
+                .poll()
+                .contains(tcp::PollState::ESTABLISHED)
+            {
+                return Err(Errno::EALREADY.into());
+            }
+
+            // the ESTABLISHED flag is set and there were no socket errors (checked above)
+            socket_ref.connect_result_is_pending = false;
+            return Ok(());
+        }
+
+        // if not an inet socket address
+        let Some(peer_addr) = peer_addr.as_inet() else {
+            return Err(Errno::EINVAL.into());
+        };
+
+        let mut peer_addr: std::net::SocketAddrV4 = (*peer_addr).into();
+
+        // https://stackoverflow.com/a/22425796
```

Comment:
A short comment about what info is expected to be at this URL would be helpful, e.g. in case the URL breaks in the future.

Reply:
Added.